provider/prometheus.go

```diff
@@ -0,0 +1,133 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+
+	"github.com/evcc-io/evcc/provider/pipeline"
+	"github.com/evcc-io/evcc/util"
+	"github.com/prometheus/client_golang/api"
+)
+
+// Prometheus provider
+type Prometheus struct {
+	log         *util.Logger
+	client      api.Client
+	clientV1Api v1.API
+	query       string
+	scale       float64
+	timeout     time.Duration
+	pipeline    *pipeline.Pipeline
+}
+
+func init() {
+	registry.Add("prometheus", NewPrometheusFromConfig)
+}
+
+func NewPrometheusFromConfig(other map[string]interface{}) (IntProvider, error) {
+	cc := struct {
+		api.Config        `mapstructure:",squash"`
+		Query             string
+		Scale             float64
+		Timeout           time.Duration
+		pipeline.Settings `mapstructure:",squash"`
+	}{
+		Scale:   1,
+		Timeout: 10 * time.Second,
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	log := util.NewLogger("prometheus")
+
+	client, err := api.NewClient(cc.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	p := NewPrometheus(log, client, cc.Query, cc.Timeout).WithScale(cc.Scale)
+
+	pipe, err := pipeline.New(cc.Settings)
+	if err == nil {
+		p = p.WithPipeline(pipe)
+	}
+
+	return p, err
+}
+
+func NewPrometheus(log *util.Logger, client api.Client, query string, timeout time.Duration) *Prometheus {
+	p := &Prometheus{
+		log:         log,
+		client:      client,
+		clientV1Api: v1.NewAPI(client),
+		query:       query,
+		scale:       1,
+		timeout:     timeout,
+	}
+
+	return p
+}
+
+// WithScale sets scaler for getters
+func (p *Prometheus) WithScale(scale float64) *Prometheus {
+	p.scale = scale
+	return p
+}
+
+// WithPipeline adds a processing pipeline
+func (p *Prometheus) WithPipeline(pipeline *pipeline.Pipeline) *Prometheus {
+	p.pipeline = pipeline
+	return p
+}
+
+func (p *Prometheus) Query() (model.Value, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
+	defer cancel()
+	res, warn, err := p.clientV1Api.Query(ctx, p.query, time.Now())
+	if err != nil {
+		return nil, err
+	}
+	p.log.TRACE.Printf("query %q: %+v", p.query, res)
+	if len(warn) > 0 {
+		p.log.WARN.Printf("query %q returned warnings: %v", p.query, warn)
+	}
+	return res, nil
+}
+
+var _ FloatProvider = (*Prometheus)(nil)
+
+// FloatGetter expects scalar value from query response as float
+func (p *Prometheus) FloatGetter() func() (float64, error) {
+	return func() (float64, error) {
+		res, err := p.Query()
+		if err != nil {
+			return 0, err
+		}
+		if res.Type() != model.ValScalar {
+			return 0, fmt.Errorf("prometheus query returned value of type %q, expected %q, consider wrapping query in scalar()", res.Type().String(), model.ValScalar.String())
```

Comment:
The logger already has name

    prometheus

so I wouldn't duplicate it in the error message.